components/p3a/uploader.h

```diff
@@ -0,0 +1,78 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_P3A_UPLOADER_H_
+#define BRAVE_COMPONENTS_P3A_UPLOADER_H_
+
+#include <memory>
+#include <string>
+
+#include "base/callback.h"
+#include "base/containers/flat_map.h"
+#include "base/memory/ref_counted.h"
+#include "brave/components/p3a/metric_log_type.h"
+
+namespace network {
+class SharedURLLoaderFactory;
+class SimpleURLLoader;
+}  // namespace network
+
+namespace net {
+class HttpResponseHeaders;
+}  // namespace net
+
+namespace p3a {
+
+constexpr char kP2AUploadType[] = "p2a";
+constexpr char kP3AUploadType[] = "p3a";
+constexpr char kP3ACreativeUploadType[] = "p3a_creative";
+
+struct P3AConfig;
+
+// Handles uploading of JSON and STAR metrics to Brave servers.
+// The endpoint used may differ depending on whether the measurement
+// is P3A, P2A, NTP-SI P3A as well as whether it is in JSON or STAR format.
+class Uploader {
+ public:
+  using UploadCompleteCallback =
+      base::RepeatingCallback<void(bool is_ok,
+                                   int response_code,
+                                   bool is_star,
+                                   MetricLogType log_type)>;
+
+  Uploader(scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+           UploadCompleteCallback upload_callback,
+           P3AConfig* config);
```

Comment:
i would simply pass it by value. If not, this should be a constant pointer that uses raw_ptr

Reply:
i mean, by const reference and store by value